test/other/test_fs_icase.cpp

```diff
@@ -0,0 +1,193 @@
+/*
+ * Copyright 2022 The Emscripten Authors.  All rights reserved.
+ * Emscripten is available under two separate licenses, the MIT license and the
+ * University of Illinois/NCSA Open Source License.  Both these licenses can be
+ * found in the LICENSE file.
+ */
+
+#include <assert.h>
+#include <dirent.h>
+#include <emscripten/emscripten.h>
+#include <stdio.h>
+#include <string.h>
+#include <string>
+#include <sys/stat.h>
+#include <unistd.h>
+#include <vector>
+
+// TODO: check different WasmFS backends.
+// TODO: check ignore case backend mounted under a different backend.
+
+void write_file(const char* fname) {
+  FILE* fp = fopen(fname, "wt");
+  assert(fp);
+  char data[] = "test";
+  printf("Write '%s' to '%s'\n", data, fname);
+  assert(fputs(data, fp) >= 0);
+  assert(fclose(fp) == 0);
+}
+
+void read_file(const char* fname) {
+  FILE* fp = fopen(fname, "rt");
+  assert(fp);
+  char buffer[10] = {};
+  assert(fgets(buffer, sizeof(buffer), fp) != NULL);
+  printf("Read '%s' from '%s'\n", buffer, fname);
+  assert(strcmp(buffer, "test") == 0);
+  assert(fclose(fp) == 0);
+}
+
+int exists(const char* fname) {
+  struct stat st = {};
+  return stat(fname, &st) == 0 ? 1 : 0;
+}
+
+std::vector<std::string> readdir(const char* dname) {
+  static std::vector<std::string> skip(
+    {".", "..", "dev", "tmp", "home", "proc"});
+  printf("Files in '%s': ", dname);
+  std::vector<std::string> files;
+  DIR* d = opendir(dname);
+  if (d) {
+    struct dirent* dir;
+    while ((dir = readdir(d)) != NULL) {
+      if (dir->d_type & (DT_DIR | DT_REG)) {
+        if (std::find(skip.begin(), skip.end(), dir->d_name) == skip.end()) {
+          files.emplace_back(dir->d_name);
+          printf("%s ", dir->d_name);
+        }
+      }
+    }
+    closedir(d);
+  }
+  printf("\n");
+  return files;
+}
+
+int main() {
+  // Create a file.
+  write_file("test.txt");
+
+  // Read and check the file.
+  struct stat st;
+  assert(stat("test.txt", &st) == 0);
+  assert(st.st_size == 4);
+  assert(exists("test.TXT"));
+  assert(exists("Test.Txt"));
+  read_file("Test.txt");
+
+  // Rename the file.
+  assert(rename("tesT.Txt", "test2.txt") == 0);
+  assert(exists("test2.txt"));
+  assert(exists("Test2.txt"));
+  read_file("Test2.txt");
+
+  // Create a directory.
+  assert(mkdir("Subdir", S_IRWXUGO) == 0);
+  assert(exists("Subdir"));
+  assert(exists("subdir"));
+  assert(mkdir("SUBDIR", S_IRWXUGO) != 0);
+  assert(errno == EEXIST);
+
+  // Check parent directory contents.
+  auto contents = readdir(".");
+  assert(contents.size() == 2);
+  assert(std::find(contents.begin(), contents.end(), "test2.txt") !=
+         contents.end());
+#ifdef WASMFS
+  assert(std::find(contents.begin(), contents.end(), "subdir") !=
+         contents.end());
+#else
+  assert(std::find(contents.begin(), contents.end(), "Subdir") !=
```

Comment:
It seems test_other and test_browser have a slightly different behavior. Because, if I make this test to run in browser, this assert fails:
`+/report_result?abort:Assertion failed: std::find(contents.begin(), contents.end(), "Subdir") != contents.end(), at: .../emscripten/test/other/test_fs_icase.cpp,102,main`

Reply:
Turns out, I was just running with default JS filesystem in browser. So it's slightly different from what I originally thought.
But anyway, JS filesystem returns original names preserving cases which can be seen in this tests (`ifdef WASMFS`).

Reply:
Makes sense, thanks. I think we could add code to `IgnoreCaseDirectory::getEntries` to map back to the original names, but I'm fine leaving that for a later PR if you prefer.

Reply:
I'd prefer to do that in a separate PR.